query.go

```diff
@@ -247,20 +247,22 @@ func (q *query) run() {
 			q.updateState(pathCtx, update)
 			cause = update.cause
 		case <-pathCtx.Done():
-			q.terminate(pathCtx, LookupCancelled)
+			q.terminate(pathCtx, cancelPath, LookupCancelled)
 		}
 
 		// termination is triggered on end-of-lookup conditions or starvation of unused peers
 		if ready, reason := q.isReadyToTerminate(); ready {
-			q.terminate(pathCtx, reason)
+			q.terminate(pathCtx, cancelPath, reason)
 		}
 
 		if q.terminated {
+			// don't wait for outstanding queries to complete.
+			return
 			// exit once all goroutines have been cleaned up
-			if q.queryPeers.NumWaiting() == 0 {
-				return
-			}
-			continue
+			// if q.queryPeers.NumWaiting() == 0 {
+			// 	return
+			// }
+			// continue
```

Comment:
If we don't wait for outstanding queries to complete then we're changing the contract. Performing this change as is will allow for writing to closed channels since some of the queryFns have channels that are closed once the lookup is completed.

Shouldn't doing the context cancellation be enough here?